openff/toolkit/utils/rdkit_wrapper.py

```diff
@@ -1102,25 +1102,29 @@ def _elf_compute_electrostatic_energy(
         ).reshape(-1, 1)
 
         # Build an exclusion list for 1-2 and 1-3 interactions.
-        excluded_pairs = {
-            *[(bond.atom1_index, bond.atom2_index) for bond in molecule.bonds],
-            *[
-                (angle[0].molecule_atom_index, angle[-1].molecule_atom_index)
-                for angle in molecule.angles
-            ],
-        }
+        excluded_x, excluded_y = zip(
+            *{
+                *[(bond.atom1_index, bond.atom2_index) for bond in molecule.bonds],
+                *[
+                    (angle[0].molecule_atom_index, angle[-1].molecule_atom_index)
+                    for angle in molecule.angles
+                ],
+            }
+        )
 
         # Build the distance matrix between all pairs of atoms.
         coordinates = conformer.value_in_unit(unit.angstrom)
 
-        distances = np.sqrt(
-            np.sum(np.square(coordinates)[:, np.newaxis, :], axis=2)
-            - 2 * coordinates.dot(coordinates.T)
-            + np.sum(np.square(coordinates), axis=1)
-        )
+        # (a - b)^2 = a^2 - 2ab + b^2
+        ab = coordinates.dot(coordinates.T)
+        # np.einsum is both faster than np.diag, and not read-only
+        diag = np.einsum("ii->i", ab)
+        # modifying in-place lets us take advantage of diag
+        ab += ab - diag - diag[..., np.newaxis]
```

Comment:
I don't follow this comment - is the point that `2ab` is quicker to compute by `einsum` and an in-place modification than doing a new dot product?

Reply:
Sorry, my notation could have been much clearer. In the `(a - b)^2 = a^2 - 2ab + b^2` method,

```python
ab = coordinates.dot(coordinates.T)
a_squared = b_squared.T = diag = np.diag(ab) = np.einsum('ii->i', ab)

distances_squared = a_squared - 2 * ab + b_squared
```

I computed the negative distances_squared inplace with this line:

```
ab += ab - diag - diag[..., np.newaxis]
```

I didn't want to allot new arrays in memory because that is one of numpy's slower actions. At this point it's just micro-optimization though, so if you think the increased readability is worth it, happy to change. This is also the part of the code I'm most iffy about. Writeability was introduced in 1.10.0 so a while ago, but I would think that `np.fill_diagonal` is more future proof than relying on this aspect of `einsum` views.

```python
def new_coords(coordinates):
    ab = coordinates.dot(coordinates.T)
    diag = np.einsum('ii->i', ab) # faster than np.diag
    ab += (ab - diag - diag[..., np.newaxis])
    diag[:] = -0.  # changing the diagonal of ab inplace
    matrix = np.sqrt(-ab)
    return matrix

def new_coords_not_inplace(coordinates):
    ab = coordinates.dot(coordinates.T)
    diag = np.einsum('ii->i', ab) # faster than np.diag
    distances_squared = diag - (2 * ab) + diag[..., np.newaxis]  # new array
    # diag is a view onto ab, not distances_squared, so we can't just set that to 0
    np.fill_diagonal(distances_squared, 0.0)
    matrix = np.sqrt(distances_squared)
    return matrix

>>> %timeit new_coords(conformer)
8.23 µs ± 53.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

>>> %timeit new_coords_not_inplace(conformer)
9 µs ± 129 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
... and none of that answered the original comment, which was that the *in-code* comment was ambiguous. The "take advantage" part of `# modifying in-place lets us take advantage of diag` refers to line 1126 (`diag[:] = -0.0`) where I can just modify the view instead of needing `np.fill_diagonal`. I put the comment there because I wanted to explain why I was computing the negative distances_squared inplace, which is an otherwise weird thing to do.

Reply:
I think my difficulty reading it is more my lack of experience doing manipulations like these - I don't think rewriting the operations is worth it, but feel free to modify comments if you think that would help future developers/future you reading this.

Re: reliability/future-proofing - I'm less worried, NumPy seems to be slow-moving these days and I'd expect any breakage to be loud and/or slower to happen than other things in the comp chem stack